Add code review guidelines for changing running infrastructure (#516)
2i2c-org/pilot-hubs: docs/contributing/code-review.md

   :::{admonition} Guidelines for reviewing changes to existing infrastructure
   :class: tip
   Focus is on both making sure the change is minimally disruptive, and on the quality of the infrastructure design.
   :::
1. **Deploy the change**. Once approved, leave a comment in the PR that you will start deploying the changes. Then `terraform apply` your change and verify it works ok.
1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:

   1. Amend your PR with changes
   2. Deploy with `terraform apply`
   3. Comment that you've made a deployment
1. **Communicate that you've finished deploying**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
1. **Self-merge or request another review**. If you didn't need to make a change to the PR, merge the PR. If you made changes to the PR, use your judgement to decide if you should request another review or if you should self-merge it directly. If you merge, just leave a comment about changes made since the initial approval.

> 

Reviewer: ```suggestion
```

However, it comes with a few caveats:

1. Locally and iteratively testing terraform is impossible - as we
   develop it, the infrastructure is deployed / modified / destroyed
   as we go.
2. Behavior of terraform code differs based on the current
   state of the infrastructure (as represented in the
   [state file](https://www.terraform.io/docs/language/state/index.html)).
3. While `terraform plan` can tell us what terraform *thinks* it is
   going to do, it might not always succeed. Permission errors or
   organizational restrictions can cause it to fail. Timeouts can happen.
   Runtime errors pop up.

Thanks to all these, it is impossible to either:

- Accurately gauge the impact of a terraform code change just from looking at the text
- Automatically deploy it via continuous deployment

As a result, we have some workflows that are specific to using Terraform, described below.

### Changes to running infrastructure

In this scenario, there is already-running infrastructure and you are deploying changes to it.

> For modifications to *existing infrastructure*, the suggested workflow is:
> 
> 
> 3. 
> 4. After the change has been successfully performed with `terraform apply`,
>    self-merge your PR.
> 
> 

Reviewer 2: Are these lines to be deleted in favor of the `To deploy changes, follow these steps:` section just below?

```suggestion
```

To deploy changes, follow these steps:

1. **Propose changes**. Change the codebase and open a PR. **Do not deploy changes yet**. Use `terraform plan` rather than `terraform apply` do understand the change before deploying it.
1. **Post the output of `terraform plan`**. In your PR's top comment, post the output of `terraform plan` whenever you update the PR. This helps others review the likely impact on our infrastructure.
1. **Wait for review and approval**. Leave the PR open for other team members to review and approve.

   :::{admonition} Guidelines for reviewing changes to existing infrastructure
   :class: tip
   Focus is on both making sure the change is minimally disruptive, and on the quality of the infrastructure design.
   :::

> 1. **Deploy the change**. Once approved, `terraform apply` your change, and make sure it works ok.
> 1. **Communicate that you've deployed**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
> 1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:
>    
>    1. Amend your PR with changes
>    2. Deploy with `terraform apply`
>    3. Comment that you've made a deployment
> 1. **Self-merge after approval**. Once someone has approved, and you are satisfied with the resulting behavior in our infrastructure, merge the PR.

Reviewer 2: I made a pass on the last four root-level bullet points. Feel free to edit my suggestion with tweaks to it as well. I wanted to make us note that we are both starting and stopping a deployment, as well as a clarification on what to do if you needed to make a changes to make it deploy successfully or not.

```suggestion
1. **Deploy the change**. Once approved, leave a comment in the PR that you will start deploying the changes. Then `terraform apply` your change and verify it works ok.
1. **Iterate as needed**. If things break or you need to iterate on the deployment, follow this cycle until the infrastructure is in the state you wish:

   1. Amend your PR with changes
   2. Deploy with `terraform apply`
   3. Comment that you've made a deployment
1. **Communicate that you've deployed**. Leave a comment in the PR that the changes have been deployed. You should always communicate when a deploy has been made!
1. **Self-merge or request another review**. If you didn't need to make a change to the PR, merge the PR. If you made changes to the PR, use your judgement to decide if you should request another review or if you should self-merge it directly and just leaving a comment about the change since it was initially approved.
```

# Code review guidelines

This document codifies our guidelines for doing code review
and merging pull requests.

> A Foolish Consistency is the Hobgoblin of Little Minds

> > - PEP 8

Reviewer 3: So the idea of this main section is to add some general guidelines and then we go with specifics such as the terraform-related stuff, right?

Author: Yep! I'm hoping we can add other guidelines soon. At the least, we should have something for:

1. doc changes
2. New hub creations
3. Hub config changes
4. ???

Reviewer 3: Those additional sections make sense to me. Not sure what else could be different from the general guidelines to deserve a specific section.

## Terraform Changes

[Terraform](https://www.terraform.io/) lets us treat our infrastructure
as code, so we can use regular code quality practices (code review,
linters, extensive documentation, etc) to keep the quality of our
infrastructure high. However, it comes with a few caveats:

1. Locally and iteratively testing terraform is impossible - as we
   develop it, the infrastructure is deployed / modified / destroyed
   as we go.
2. Behavior of a bit of terraform code differs based on the current
   state of the infrastructure (as represented in the
   [state file](https://www.terraform.io/docs/language/state/index.html)).
3. While `terraform plan` can tell us what terraform *thinks* it is
   going to do, it might not always succeed. Permission errors or
   organizational restrictions can cause it to fail. Timeouts can happen.

>    Runtime errors pop up.

Reviewer 3: I like this section explaining the associated complexities!

Author: yay!

Thanks to all these, it is impossible to accurately gauge the impact of
a terraform code change just from looking at the text nor automatically
deploy it via continuous deployment.

The guiding principles for terraform code review are:

1. Person who writes the code also does the `terraform apply`.

> 2. Terraform code merged into the repo must already be successfully applied.

Reviewer: Can we add a step or checkbox like "explicitly mark when `terraform apply` has been run"?

Reviewer 3: Makes sense to me!

3. For new infrastructure, focus of code review is just on infrastructure design review.
4. For existing infrastructure, focus is on both making sure the change
   is minimally disruptive, and on infrastructure design.

### When setting up new infrastructure

For creating *new* infrastructure, the suggested workflow is:

1. Iteratively develop the code locally, testing it with `terraform apply`
   as you go.
2. Start a PR early, and solicit feedback. The primary goal here is to
   do a review of the *infrastructure design*, rather than of any particular
   `terraform plan` output.
3. Once someone else approves your PR, you must:

   1. Destroy the current infra you have setup with `terraform destroy`
   2. `terraform apply` from scratch, so new infra is setup from scratch
      based on the reviewed state of the PR. If things break, amend your
	  PR until it unbreaks. This might solicit adding another cycle of
	  code review.

>    3. Self-merge the PR once the infra works as you would like.

Reviewer: Given that the person runs `terraform apply` on their own, what does the merge step do? Simply make the codebase reflect what has been applied? If so, then (related to my previous comment), there will be an in-between time when the state of the cluster is different from the state of `main`. What can we do to ensure that this difference is very short, or at least that it is clear to people?

Reviewer 3: I don't think there is a way to enforce/ensure that... so we need policy. Maybe we need to be more strong and establish a max amount of time the difference could live??

Reviewer 4: I agree that this is difficult to enforce, but I think `terraform apply` should be run immediately after the merge, by whomever merged.

Author: > but I think terraform apply should be run immediately after the merge, by whomever merged.

The options are wether it is run before merge or after merge. I personally prefer before merge, since it allows for iteration when `terraform plan` fails (as it often does). If we were to run it after merge, it could break and we'll end up with terraform code in our main branch that fails...
